driver/src/orderbook/streamed/updating_orderbook.rs

```diff
@@ -0,0 +1,75 @@
+use super::*;
+use crate::contracts::stablex_contract::batch_exchange;
+use anyhow::{anyhow, bail, Result};
+use block_timestamp::BlockTimestamp;
+use ethcontract::{contract::Event, errors::ExecutionError};
+use futures::stream::{BoxStream, StreamExt as _};
+use orderbook::Orderbook;
+use std::future::Future;
+use std::sync::{mpsc, Arc, RwLock};
+
+// TODO:
+// `pub struct UpdatingOrderbook` that also owns an `Arc<RwLock<Orderbook>>` and creates a
+// background thread running `UpdatingOrderbookThread::update_with_events`.
+// This way `UpdatingOrderbookThread` is independently testable.
+
+/// Update the orderbook with events from the stream forever or until exit_indicator is dropped.
+///
+/// Returns Ok when exit_indicator is dropped.
+/// Returns Err if the stream ends.
+async fn update_with_events_forever(
+    orderbook: Arc<RwLock<Orderbook>>,
+    mut block_timestamp: impl BlockTimestamp,
+    exit_indicator: mpsc::Receiver<()>,
+    past_events: impl Future<Output = Result<Vec<Event<batch_exchange::Event>>, ExecutionError>>,
```

Comment:
This might make more sense to be a stream since past events will come in pages. Alternatively the `ethcontract` implementation can collect it into a `Vec` for you as well.

Reply:
Now that I think about it, collecting in a `Vec` will be way easier to implement, so lets leave it as is.

Reply:
I don't mind either way. I used these types because that's what I prototyped in that other pr in `StablexContract`.